mlir-sycl/lib/Conversion/SYCLToSPIRV/SYCLToSPIRV.cpp

```diff
@@ -0,0 +1,233 @@
+  const auto loc = op->getLoc();
+  const auto indexTy = rewriter.getIndexType();
+  const auto alloca = static_cast<Value>(rewriter.create<memref::AllocaOp>(
+      loc, MemRefType::get({dimensions}, indexTy)));
+  for (int64_t i = 0; i < dimensions; ++i) {
+    const auto c =
+        static_cast<Value>(rewriter.create<arith::ConstantIndexOp>(loc, i));
+    const auto val = static_cast<Value>(rewriter.create<arith::IndexCastOp>(
+        loc, indexTy, getDimension(rewriter, loc, values, i)));
+    rewriter.create<AffineStoreOp>(loc, val, alloca, c);
+  }
+  index = rewriter.create<arith::IndexCastOp>(loc, indexTy, index);
+  rewriter.replaceOpWithNewOp<AffineLoadOp>(op, alloca, index);
+}
+
+/// Converts n-dimensional operations of type \tparam OpTy not being passed an
+/// argument to a call to a SPIRV builtin.
+template <typename OpTy> class GridOpPatternIndex : public GridOpPattern<OpTy> {
+public:
+  using GridOpPattern<OpTy>::GridOpPattern;
+
+  LogicalResult match(OpTy op) const override {
+    return success(op.getNumOperands() == 1);
+  }
+
+  void rewrite(OpTy op, typename OpTy::Adaptor opAdaptor,
+               ConversionPatternRewriter &rewriter) const final {
+    rewriteNDIndex(
+        op, GridOpPattern<OpTy>::spirvBuiltin, opAdaptor.getDimension(),
+        *this->template getTypeConverter<SPIRVTypeConverter>(), rewriter);
+  }
+};
+
+/// Replace \p op with a sequence of operations that:
+/// 1. Allocate a new object of the result type in the stack;
+/// 2. Load the object;
+/// 3. Initialize the dimensions of the object with the values in builtin \p
+/// builtin.
+void rewriteNDNoIndex(Operation *op, spirv::BuiltIn builtin,
+                      const SPIRVTypeConverter &typeConverter,
+                      ConversionPatternRewriter &rewriter) {
+  // This conversion is platform dependent
+  const auto dimensions = getDimensions(op->getResultTypes()[0]);
+  const auto values = spirv::getBuiltinVariableValue(
+      op, builtin, typeConverter.getIndexType(), rewriter);
+  const auto loc = op->getLoc();
+  const auto getIndexTy = rewriter.getIntegerType(32);
+  const auto targetIndexType = rewriter.getI64Type();
+  const auto dimMtTy = MemRefType::get(dimensions, targetIndexType, {}, 4);
+  // Allocate
+  const auto resTy = op->getResultTypes()[0];
+  const auto alloca = static_cast<Value>(
+      rewriter.create<memref::AllocaOp>(loc, MemRefType::get(1, resTy)));
+  // Load
+  const auto zero =
+      static_cast<Value>(rewriter.create<arith::ConstantIndexOp>(loc, 0));
+  const auto res = static_cast<Value>(
+      rewriter.replaceOpWithNewOp<AffineLoadOp>(op, alloca, zero));
+  const auto argumentTypes =
+      rewriter.getTypeArrayAttr({MemRefType::get(1, resTy, {}, 4), getIndexTy});
+  const auto functionName = rewriter.getAttr<FlatSymbolRefAttr>("operator[]");
+  // Initialize
+  for (int64_t i = 0; i < dimensions; ++i) {
+    const auto index = static_cast<Value>(
```

Comment:
Propose to `drop static_cast` here and elsewhere and declare variable with `Value` type explicitly.